### PR TITLE
Plugin interface for signing. 

### PR DESCRIPTION
**Pull request by gokarnm**
Related issue notaryproject/roadmap/issues/26.

* Includes plugin interfaces for signing workflow
* Addresses feedback on plugin mechanism and versioning
* Added validations performed by Notation for plugin responses
* Added error contract and error codes
* [hackmd doc](https://hackmd.io/jhYbA4aNRFW7nmp7gPTIBA?view)

Signed-off-by: Milind Gokarn <gokarnm@amazon.com>

### CODE REVIEW

**Comment by priteshbandi** [comment]
Can you please run markdown linter on the change.

File: specs/plugin-extensibility.md
Comment (rchincha) [inline specs/plugin-extensibility.md:30]: If there are multiple plugins in the plugins-dir, does `com.amazonaws.signer.nv2plugin` map to a particular plugin?

File: specs/plugin-extensibility.md
Comment (rchincha) [inline specs/plugin-extensibility.md]: Nitpicking.

`com.amazonaws.signer.nv2plugin` repetitive in `~/.notation/plugins/com.amazonaws.signer.nv2plugin/notation-com.amazonaws.signer.nv2plugin

File: specs/plugin-extensibility.md
Comment (rchincha) [inline specs/plugin-extensibility.md]: `request.key` must be optional since signing keys may be private to the remote API.
You are really asking for two modes here:
1. I have the key but can't do crypto (at least not native to notation), so let plugin handle that
2. I don't the key and of course can't do crypto, so let plugin ship the payload and get it signed.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: That was the intent, `/plugin` has a child dir called `{plugin-name}`, which contains an executable called `notation-{plugin-name}`.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: The `request.key` is the key definition in notation config, here `name` is a friendly name to refer from CLI, `id` is the plugin/crypto specific identifier that allows plugin to know which key to use. They do not contain the value of signing key itself. I should probably change `request.key` to `request.keyDefinition` or similar to make this clearer.

``` json
"keys": [
      {  
        "name": "mysigningkey",  
        "id" : "keyid",
        "plugin": "com.example.nv2plugin"
      }  
```

**Comment by SteveLasker** [comment]
@aramase, can you help here?

File: specs/plugin-extensibility.md
Comment (rchincha) [reply specs/plugin-extensibility.md]: `~/.notation/plugins/com.amazonaws.signer.nv2plugin/notation-com.amazonaws.signer.nv2plugin` => `~/.notation/plugins/com.amazonaws.signer/nv2`


Your call though.

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md]:
```suggestion
Notation will invoke plugins as executable, pass parameters using command line arguments, and use standard IO streams to pass request/response payloads. This mechanism is used as Go language (used to develop [Notation library](https://github.com/notaryproject/notation-go-lib)) does not have a [in built support](https://github.com/golang/go/issues/19282) to load and execute plugins that works across OS platforms. Other mechanisms like gRPC require every plugin to be implemented as a service/daemon.
```

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md:27]: >Plugins intended for public distribution should also include instructions for users to verify the authenticity of the plugin.

Should we say MUST instead of SHOULD include instructions for users to verify the authenticity of the plugin?
If a user cannot determine the authenticity of the plugin then they cannot trust plugin.

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md:28]: nit: The `notation` in `notation-{plugin-name}` of `~/.notation/plugins/{plugin-name}/notation-{plugin-name}` is redundant.

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md]: > The request, response and error contract for each command will be defined.

Haven't we already defined this?

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md]: What happens if the exit code is neither 0 nor 1? Should implementation throw an error or consider all non-zero exit codes as an error?

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md:67]: Should returning error JSON be optional and notation should consider it as InternalFailure?
There are cases where the plugin might fail unexpectedly and might not be able to emit error JSON. Or the user might want to implement a plugin with minimal error handling.

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md]: nit: this should be an array.

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md:108]: > Currently this list must include only one version, per major version.

Does this mean plugin publishers should only specify a major version of the contract?

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md]: Missing hyperlink for `[Certificate Requirements]`?

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md:196]: s/ERROR/INTERNAL_ERROR ?

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md:195]: Shouldn't this be handled internally by the plugin? Why does notation need to know about access denied, timeout and throttling?

File: specs/plugin-extensibility.md
Comment (priteshbandi) [inline specs/plugin-extensibility.md]: Will Notation generate timestamp signature for `Signature Envelope Generator` plugin or its responsibility of plugin publisher?

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: Was waiting for your PR merge :), will update

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md:27]: There is limited things the spec can do here, unless there is a uniform mechanism to enforce it and Notation to verify. Plugins packages may be distributed for internal or public consumption and may be signed/unsigned or may use any mechanism to assert authenticity.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: Reworded.
```suggestion
* The command will be passed as the first argument to the plugin e.g. `notary-{plugin-name} <command>`. A JSON request is passed using `stdin`. The plugin is expected to return a JSON response through `stdout` with a `0` exit code for successful response, and a non-zero exit code with a JSON error response in `stderr` for error response. Each command defines its request, response and error contract. To avoid any additional content like debug or info level logging from dependencies and inbuilt libraries, the plugin implementation should redirect any output to `stdout` on initialization, and only send the JSON response away from `stdout` when the command execution completes. E.g. For golang, set [`os.Stdout`](https://pkg.go.dev/os#pkg-variables) to point to a log file.
```

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: Good point, addressed.
```suggestion
* For an error response, every command returns a non-zero exit code 1, with an OPTIONAL JSON error response in `stderr`. It is recommended to return an error response to help user troubleshoot the error.  There is no need to send different exit codes for different error conditions, as Notation (which will call the plugin and parse the response) will use `error-code` in the error response to interpret different error conditions if needed. Notation will attempt to parse the error response in `stderr` when exit code is 1, else treat it as a general error for any other non-zero exit codes. An implementation can 
```

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md:67]: Addressed for previous comment.

File: specs/plugin-extensibility.md
Comment (gokarnm) [inline specs/plugin-extensibility.md]:
```suggestion
  // List of contract versions supported by the plugin, one per major version
  "supported-contract-versions" : [ ],
```

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]:
```suggestion
           3. Check that the `response.certificateChain` conforms to [Certificate Requirements](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#certificate-requirements).
```

File: specs/plugin-extensibility.md
Comment (sajayantony) [inline specs/plugin-extensibility.md]: Having a issue to track this before the next release would be good. Similarly for the **Threat Model** as well.

**Comment by sajayantony** [comment]
If we create the required issues for TBD section would be good for initial Push.

**Review by sajayantony — APPROVED** [review]
Good for initial push with issues created.

File: specs/plugin-extensibility.md
Comment (aramase) [inline specs/plugin-extensibility.md]:
```suggestion
* Every plugin MUST implement a metadata discovery command called `discover`. 
```

File: specs/plugin-extensibility.md
Comment (aramase) [inline specs/plugin-extensibility.md:108]: Do we also need to define the supportability contract for major and minor versions? How long will a particular minor version be supported in notation when there are multiple minor versions?

File: specs/plugin-extensibility.md
Comment (aramase) [inline specs/plugin-extensibility.md]:
```suggestion
           3. Check that the `response.certificateChain` conforms to [Certificate Requirements].
```

File: specs/plugin-extensibility.md
Comment (shizhMSFT) [inline specs/plugin-extensibility.md]:
```suggestion
Notation will invoke plugins as executable, pass parameters using command line arguments, and use standard IO streams to pass request/response payloads. This mechanism is used as Go language (used to develop [Notation library](https://github.com/notaryproject/notation-go-lib)) does not have a [built-in support](https://github.com/golang/go/issues/19282) to load and execute plugins that works across OS platforms. Other mechanisms like gRPC require every plugin to be implemented as a service/deamon.
```

File: specs/plugin-extensibility.md
Comment (aramase) [inline specs/plugin-extensibility.md:178]: Should we use `key` here instead of `keyId` as the key definition is passed as part of the request?

File: specs/plugin-extensibility.md
Comment (aramase) [inline specs/plugin-extensibility.md:196]: These error codes are good to be surfaced out but IMO notation should not perform any retry action. In case of throttled requests, it's possible request can only be retried after a certain time defined in `retry-after` header. The plugin can decide to (1) retry the request internally before giving up and returning an error or (2) return the `retry-after` header info as part of error message, so the user can retry the request after the time has elapsed.

Comment (aramase) [reply specs/plugin-extensibility.md:196]: This also makes me think there should be a context defined with timeout as part of the request to the plugin. We can use [CommandContext](https://pkg.go.dev/os/exec#CommandContext) to ensure the request doesn't hang indefinitely

File: specs/plugin-extensibility.md
Comment (shizhMSFT) [reply specs/plugin-extensibility.md:28]: Similar pattern for executables can be found in [application bundles](https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFBundles/BundleTypes/BundleTypes.html).

File: specs/plugin-extensibility.md
Comment (aramase) [inline specs/plugin-extensibility.md:99]: If `SIGNATURE_GENERATOR` and `SIGNATURE_ENVELOPE_GENERATOR` are supported by the plugin which one takes precedence when `notation sign` is invoked?

File: specs/plugin-extensibility.md
Comment (shizhMSFT) [inline specs/plugin-extensibility.md]: nit: `jsonc` instead of `json` renders better.

```jsonc
{
  // Each command defines expected error codes
  "error-code" : "<error code>",  
  // Plugin defined error message.
  "error-message" : "User friendly error message"
  // Optional plugin defined additional key value pairs related to the error.
  "error-metadata" : {
    "key1" : "value1",
    "key2" : "value2"
  }
}

File: specs/plugin-extensibility.md
Comment (shizhMSFT) [inline specs/plugin-extensibility.md]: Could you provide an example request?

File: specs/plugin-extensibility.md
Comment (shizhMSFT) [inline specs/plugin-extensibility.md]:
```suggestion
        1. Execute the plugin with `generate-envelope` command, set `request.key` to key definition (from `/notation/config.json`), `request.payload` to base64 encoded descriptor, `request.payloadType` to `application/vnd.oci.descriptor.v1+json` and `request.signatureEnvelopeType` to a pre-defined type (default to `application/vnd.cncf.notary.v2.jws.v1`).
```

**Review by shizhMSFT — APPROVED** [review]
LGTM on this initial spec. We can add the specs for timestamping and verification in future PRs.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: I'd prefer to keep the existing convention, I agree it's repetitive, though it does not directly impact the end user experience. Initially the executable name was `{plugin-name}`, then we prefixed notation following docker's convention for CLI plugins. @sajayantony let me know if you have any feedback to simplify this.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md:99]: Currently we expect a signing plugin to implement one of the signing interfaces not both. Let me know if this does not work for a specific case.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md:108]: @priteshbandi @aramase, plugin publishers specify the major.minor version of contract interface, for each major version. We are only putting in place mechanisms to support breaking and non breaking changes to plugin interface. These will not rev at the same rate as Notation updates. Notation should ideally support all minor versions of contract. The intent is plugin publishers can choose to implement the latest version of contract, but we don't want to break support for plugins implementing older minor versions of contract (within the major version).

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md:178]: The `key` in the request was actually the complete key definition, I'll rename that field to `keyDefinition` to be clearer.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md:196]: This is supposed to be a catch all error that would cover any known, and unhandled plugin errors that do not fall into other categories.

Comment (gokarnm) [reply specs/plugin-extensibility.md:196]: Agreed, Notation should not retry. The errors would be surfaced to caller invoking Notation, and it can decide to retry if needed.

Comment (gokarnm) [reply specs/plugin-extensibility.md:196]: Can you provide more details about the context with timeout as part of the request, maybe with an example, I didn't understand that.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: Good point, will address in next rev.

File: specs/plugin-extensibility.md
Comment (gokarnm) [reply specs/plugin-extensibility.md]: Done.

**Comment by gokarnm** [comment]
Addressed review feedback, linked open issues for verification plugin and threat model, and squashed commits.

**Review by priteshbandi — APPROVED** [review]
LGTM

**Review by NiazFK — COMMENTED** [review]
LGTM

**Review by NiazFK — APPROVED** [review]

**Review by aramase — APPROVED** [review]
lgtm

for some of the outstanding questions, we can follow up during implementation.

**Review by SteveLasker — APPROVED** [review]
LGTM

**Pull request merged**
